File: android/app/src/main/java/com/hafizulrifkihawari/jpflashcard/KotdPicker.kt
package com.hafizulrifkihawari.jpflashcard

import android.content.Context
import org.json.JSONArray
import java.io.BufferedReader
import java.io.InputStreamReader
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

/** Mirrors kotd-data.js's `example: { jp, reading, meaning }`. */
data class KotdExample(val jp: String, val reading: String, val meaning: String)

/**
 * Mirrors kotd-data.js's RAW entry shape. `single` entries use kanji/onyomi/
 * kunyomi; `compound` entries use word/reading/parts — fields not applicable
 * to a given `type` are null, matching the JS source's duck-typing rather
 * than splitting into two exclusive types.
 */
data class KotdEntry(
    val id: Int,
    val type: String,
    val kanji: String?,
    val word: String?,
    val onyomi: String?,
    val kunyomi: String?,
    val reading: String?,
    val meaning: String,
    val meaningEn: String,
    val level: String,
    val example: KotdExample,
    val parts: List<Int>?
)

/** How many entries are featured per day. Must match kotd.js's KOTD_DAILY_COUNT (3). */
const val KOTD_DAILY_COUNT = 3

private const val DAY_MS = 24L * 60 * 60 * 1000

/** Same UTC "YYYY-MM-DD" stamp as srs.js's todayStamp(). */
fun todayStampUtc(now: Instant = Instant.now()): String =
    DateTimeFormatter.ISO_LOCAL_DATE.withZone(ZoneOffset.UTC).format(now)

/**
 * Epoch-day index for a UTC midnight, matching kotd.js's
 * `Math.floor(Date.parse(stamp + "T00:00:00Z") / DAY_MS)`.
 */
fun dayIndexFor(stamp: String): Long {
    val midnightUtcMillis = Instant.parse("${stamp}T00:00:00Z").toEpochMilli()
    return Math.floorDiv(midnightUtcMillis, DAY_MS)
}

/**
 * Deterministic per-UTC-day picker — a literal port of kotd.js's pickDaily().
 * Intentionally uses plain (non floor-adjusted) `%` to mirror the JS exactly;
 * dayIndex is always positive for any real calendar date, so this doesn't
 * need JS's negative-modulo caveats. See kotd.js for the original.
 */
fun <T> pickDaily(list: List<T>, count: Int, stamp: String): List<T> {
    val n = list.size
    if (n == 0) return emptyList()
    val c = minOf(count, n)
    val dayIndex = dayIndexFor(stamp)
    val start = ((dayIndex * c) % n).toInt()
    return (0 until c).map { list[(start + it) % n] }
}

fun loadRawFromAssets(context: Context): List<KotdEntry> {
    val text = context.assets.open("kotd-data.json").use { stream ->
        BufferedReader(InputStreamReader(stream, Charsets.UTF_8)).readText()
    }
    val arr = JSONArray(text)
    val out = ArrayList<KotdEntry>(arr.length())
    for (i in 0 until arr.length()) {
        val o = arr.getJSONObject(i)
        val ex = o.getJSONObject("example")
        val parts = if (o.has("parts") && !o.isNull("parts")) {
            val partsArr = o.getJSONArray("parts")
            (0 until partsArr.length()).map { partsArr.getInt(it) }
        } else null
        fun optNullableString(key: String): String? = if (o.has(key) && !o.isNull(key)) o.getString(key) else null

        out.add(
            KotdEntry(
                id = o.getInt("id"),
                type = o.getString("type"),
                kanji = optNullableString("kanji"),
                word = optNullableString("word"),
                onyomi = optNullableString("onyomi"),
                kunyomi = optNullableString("kunyomi"),
                reading = optNullableString("reading"),
                meaning = o.getString("meaning"),
                meaningEn = o.getString("meaningEn"),
                level = o.getString("level"),
                example = KotdExample(ex.getString("jp"), ex.getString("reading"), ex.getString("meaning")),
                parts = parts
            )
        )
    }
    return out
}

fun buildIndex(list: List<KotdEntry>): Map<Int, KotdEntry> = list.associateBy { it.id }

/** Convenience for the widget: today's featured entry (index 0 of the day's picks). */
fun todaysFeaturedEntry(context: Context): KotdEntry? {
    val raw = loadRawFromAssets(context)
    val picks = pickDaily(raw, KOTD_DAILY_COUNT, todayStampUtc())
    return picks.firstOrNull()
}

File: ios/App/KotdWidgetExtension/KotdPicker.swift
import Foundation

/// Mirrors kotd-data.js's `example: { jp, reading, meaning }`.
struct KotdExample: Codable {
    let jp: String
    let reading: String
    let meaning: String
}

/// Mirrors kotd-data.js's RAW entry shape. `single` entries use kanji/onyomi/
/// kunyomi; `compound` entries use word/reading/parts — fields that don't
/// apply to a given `type` are nil, matching the JS source's duck-typing
/// rather than splitting into two exclusive types.
struct KotdEntry: Codable {
    let id: Int
    let type: String
    let kanji: String?
    let word: String?
    let onyomi: String?
    let kunyomi: String?
    let reading: String?
    let meaning: String
    let meaningEn: String
    let level: String
    let example: KotdExample
    let parts: [Int]?
}

/// How many entries are featured per day. Must match kotd.js's KOTD_DAILY_COUNT (3).
let kotdDailyCount = 3

private let dayMs: Double = 24 * 60 * 60 * 1000

/// Same UTC "YYYY-MM-DD" stamp as srs.js's todayStamp(). `en_US_POSIX` is
/// required here — without a fixed locale, DateFormatter parsing/formatting
/// can behave unexpectedly under some device locale/calendar settings (a
/// real iOS gotcha with no JS equivalent, since Date.parse isn't
/// locale-sensitive).
func todayStampUTC(_ now: Date = Date()) -> String {
    let f = DateFormatter()
    f.dateFormat = "yyyy-MM-dd"
    f.timeZone = TimeZone(identifier: "UTC")
    f.locale = Locale(identifier: "en_US_POSIX")
    return f.string(from: now)
}

private func utcStampFormatter() -> DateFormatter {
    let f = DateFormatter()
    f.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
    f.timeZone = TimeZone(identifier: "UTC")
    f.locale = Locale(identifier: "en_US_POSIX")
    return f
}

/// UTC midnight Date for a "YYYY-MM-DD" stamp.
func utcMidnight(forStamp stamp: String) -> Date {
    utcStampFormatter().date(from: "\(stamp)T00:00:00Z")!
}

/// Epoch-day index for a UTC midnight, matching kotd.js's
/// `Math.floor(Date.parse(stamp + "T00:00:00Z") / DAY_MS)`.
func dayIndex(forStamp stamp: String) -> Int {
    Int(floor(utcMidnight(forStamp: stamp).timeIntervalSince1970 * 1000 / dayMs))
}

/// Deterministic per-UTC-day picker — a literal port of kotd.js's pickDaily().
func pickDaily<T>(_ list: [T], _ count: Int, _ stamp: String) -> [T] {
    let n = list.count
    if n == 0 { return [] }
    let c = min(count, n)
    let di = dayIndex(forStamp: stamp)
    let start = (di * c) % n
    return (0..<c).map { list[(start + $0) % n] }
}

enum KotdData {
    /// Loads and decodes the bundled kotd-data.json (added as a resource of
    /// the KotdWidgetExtension target specifically).
    static func loadAll() -> [KotdEntry] {
        guard let url = Bundle.main.url(forResource: "kotd-data", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let entries = try? JSONDecoder().decode([KotdEntry].self, from: data) else {
            return []
        }
        return entries
    }

    /// Today's featured entry (index 0 of the day's picks), for a given stamp.
    static func featuredEntry(forStamp stamp: String, in all: [KotdEntry]) -> KotdEntry? {
        pickDaily(all, kotdDailyCount, stamp).first
    }
}
